packages/next/src/server/web/spec-extension/adapters/request-cookies.ts

```diff
@@ -144,29 +152,86 @@ export class MutableRequestCookiesAdapter {
               )
               try {
                 target.delete(...args)
+                return wrappedCookies
               } finally {
                 updateResponseCookies()
               }
             }
           case 'set':
-            return function (
-              ...args:
-                | [key: string, value: string, cookie?: Partial<ResponseCookie>]
-                | [options: ResponseCookie]
-            ) {
+            return function (...args: SetCookieArgs) {
               modifiedCookies.add(
                 typeof args[0] === 'string' ? args[0] : args[0].name
               )
               try {
-                return target.set(...args)
+                target.set(...args)
+                return wrappedCookies
               } finally {
                 updateResponseCookies()
               }
             }
+
           default:
             return ReflectAdapter.get(target, prop, receiver)
         }
       },
     })
+
+    return wrappedCookies
+  }
+}
+
+export function wrapWithMutableAccessCheck(
+  responseCookies: ResponseCookies
+): ResponseCookies {
+  const wrappedCookies = new Proxy(responseCookies, {
+    get(target, prop, receiver) {
+      switch (prop) {
+        case 'delete':
+          return function (...args: [string] | [ResponseCookie]) {
+            ensureCookiesAreStillMutable('cookies().delete')
+            target.delete(...args)
+            return wrappedCookies
```

Comment:
If you instead used the ReflectAdapter you wouldn't need to rewrite this here to call delete explicitly.

Reply:
sorry, i don't understand, wdym?

i'm pretty sure we need to rewrite it like this so that `cookies().set(...).set(...)` correctly operates on the proxied version (and not `target`). see the test i added here https://github.com/vercel/next.js/pull/71094/files#diff-cc4b89e37b032bad6aa2365dd3308519cf1e2de8f2668609c68c60183d503d4fR62. not sure what ReflectAdapter would change here

Reply:
discussed in DMs, code's alright, just unintuitive